src/herder/TransactionQueue.cpp

```diff
@@ -0,0 +1,269 @@
+            {"herder", "pending-txs", fmt::format("age{}", i)}));
+    }
+}
+
+TransactionQueue::AddResult
+TransactionQueue::tryAdd(TransactionFramePtr tx)
+{
+    if (isBanned(tx->getFullHash()))
+    {
+        return TransactionQueue::AddResult::STATUS_TRY_AGAIN_LATER;
+    }
+
+    if (contains(tx))
+    {
+        return TransactionQueue::AddResult::STATUS_DUPLICATE;
+    }
+
+    auto state = getAccountTransactionQueueState(tx->getSourceID());
+    LedgerTxn ltx(mApp.getLedgerTxnRoot());
+    if (!tx->checkValid(ltx, state.mMaxSeq))
+    {
+        return TransactionQueue::AddResult::STATUS_ERROR;
+    }
+
+    auto sourceAccount = stellar::loadAccount(ltx, tx->getSourceID());
+    if (getAvailableBalance(ltx.loadHeader(), sourceAccount) - tx->getFeeBid() <
+        state.mTotalFees)
+    {
+        tx->getResult().result.code(txINSUFFICIENT_BALANCE);
+        return TransactionQueue::AddResult::STATUS_ERROR;
+    }
+
+    auto map = findOrAdd(mPendingTransactions[0], tx->getSourceID());
+    map->addTx(tx);
+
+    return TransactionQueue::AddResult::STATUS_PENDING;
+}
+
+void
+TransactionQueue::remove(std::vector<TransactionFramePtr> const& dropTxs)
+{
+    for (auto& m : mPendingTransactions)
+    {
+        if (m.empty())
+        {
+            continue;
+        }
+
+        std::set<std::shared_ptr<TxMap>> toRecalculate;
+
+        for (auto const& tx : dropTxs)
+        {
+            auto const& acc = tx->getSourceID();
+            auto const& txID = tx->getFullHash();
+            auto i = m.find(acc);
+            if (i != m.end())
+            {
+                auto& txs = i->second->mTransactions;
+                auto j = txs.find(txID);
+                if (j != txs.end())
+                {
+                    txs.erase(j);
+                    if (txs.empty())
+                    {
+                        m.erase(i);
+                    }
+                    else
+                    {
+                        toRecalculate.insert(i->second);
+                    }
+                }
+            }
+        }
+
+        for (auto txm : toRecalculate)
+        {
+            txm->recalculate();
+        }
+    }
+}
+
+bool
+TransactionQueue::contains(TransactionFramePtr tx) const
+{
+    return std::any_of(std::begin(mPendingTransactions),
+                       std::end(mPendingTransactions),
+                       [&](AccountTxMap const& map) {
+                           auto txMap = map.find(tx->getSourceID());
+                           if (txMap == map.end())
+                           {
+                               return false;
+                           }
+
+                           auto& transactions = txMap->second->mTransactions;
+                           return transactions.find(tx->getFullHash()) !=
+                                  std::end(transactions);
+                       });
+}
+
+AccountTransactionsQueueState
+TransactionQueue::getAccountTransactionQueueState(
```

Comment:
`AccountTransactionQueueState` is still implying a state, I was suggesting earlier `AccountTransactionQueueInfo` or `AccountTransactionQueueSummary`